### ISSUE TITLE
Training Inception v3 from Keras TF fails with ArrayIndexOutOfBoundsException

### ISSUE THREAD

**Issue by rohrl** [deeplearning4j/deeplearning4j#2689]
Version 0.7.2 + latest model import module from master.

Running following code fails:
`ComputationGraphConfiguration config = importKerasModelConfiguration(
            "KerasInception3.json"); // from Keras Tensorflow

        ComputationGraph model = new ComputationGraph(config);
        model.init();

        double[] flattened = new double[299 * 299 * 3];
        Arrays.fill(flattened, 0.05);

        INDArray input = Nd4j.create(flattened, new int[] { 1, 3, 299, 299 }, 'c'); 

        double[] outputArray = new double[1000];
        Arrays.fill(outputArray, 0.01);

        INDArray output = Nd4j.create(outputArray);
        
        model.fit(new INDArray[] { input }, new INDArray[] { output });`

Fails with:
`Caused by: java.lang.ArrayIndexOutOfBoundsException: 0
	at org.deeplearning4j.nn.graph.ComputationGraph.calcBackpropGradients(ComputationGraph.java:1198)
	at org.deeplearning4j.nn.graph.ComputationGraph.computeGradientAndScore(ComputationGraph.java:969)
	at org.deeplearning4j.optimize.solvers.BaseOptimizer.gradientAndScore(BaseOptimizer.java:151)
	at org.deeplearning4j.optimize.solvers.StochasticGradientDescent.optimize(StochasticGradientDescent.java:54)
	at org.deeplearning4j.optimize.Solver.optimize(Solver.java:51)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:833)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:798)
	at deeplearning.Main.main(Main.java:71)
	... 5 more`

**Comment by AlexDBlack** [comment]
Inception v3 has 2 softmax outputs, right? You are provinig a single label array.

Check model.getNumOutputArrays() to be sure, and also try model.fit(new INDArray[] { input }, new INDArray[] { output, output });

**Comment by rohrl** [comment]
model.getNumOutputArrays() = 1.

I tried "{output, output}" , it gives:  
_java.lang.IllegalArgumentException: Invalid output array: network has 1 outputs, but array is of length 2_

The network comes from Keras:
https://github.com/fchollet/deep-learning-models/releases/download/v0.2/inception_v3_weights_tf_dim_ordering_tf_kernels.h5
I exported it to json using Keras model.to_json

**Comment by AlexDBlack** [comment]
OK, thanks for checking. I think you'll need @eraly or @turambar to comment on that one.

**Comment by turambar** [comment]
@rohrl On our radar. Some major changes to Keras model import are coming, including some important fixes for support for Convolution layers (Theano and TF backends). Once those are done, we'll take a look at this.

**Comment by rohrl** [comment]
Could you give some estimate on when all these fixes will be released?

**Comment by agibsonccc** [comment]
@rohrl we don't comment on that publicly. This is to avoid misleading people about dates of when to expect things. Skymind customers get these notices as we can safely prioritize them since we manage point releases for them. Open source releases are usually prone to *alot* of variables. 

I would provide estimates for open source if we had snapshots but we don't have those either due to the need to build for windows,mac,linux (multiple chips), and android (multiple chips)

**Comment by turambar** [comment]
Fixed in master and in upcoming release.

**Comment by rohrl** [comment]
I'm still getting this same error when trying to train a Inception V3 imported from Keras (TH this time). Could you please double check if the fix works correctly? 
Maybe I'm doing something wrong.. here's the piece of code I run:
```
        ParentPathLabelGenerator labelMaker = new ParentPathLabelGenerator();
        File mainPath = new File(dataRootPath);
        FileSplit fileSplit = new FileSplit(mainPath, NativeImageLoader.ALLOWED_FORMATS, rng);
        BalancedPathFilter pathFilter = new BalancedPathFilter(rng, labelMaker, numExamples, numLabels, batchSize);

        InputSplit[] inputSplit = fileSplit.sample(pathFilter, numExamples * (1 + splitTrainTest), numExamples * (1 - splitTrainTest));
        InputSplit trainData = inputSplit[0];
        InputSplit testData = inputSplit[1];

        DataNormalization scaler = new ImagePreProcessingScaler(0, 1);

        network.init();

        ImageRecordReader recordReader = new ImageRecordReader(height, width, channels, labelMaker);
        DataSetIterator dataIter;
        MultipleEpochsIterator trainIter;

        log.info("Train model....");

        recordReader.initialize(trainData, null);
        dataIter = new RecordReaderDataSetIterator(recordReader, batchSize, 1, numLabels);
        scaler.fit(dataIter);
        dataIter.setPreProcessor(scaler);
        trainIter = new MultipleEpochsIterator(epochs, dataIter, nCores);
        
        network.fit(trainIter);
```
It's taken from tutorial on training VGG16, I just replaced it with InceptionV3

**Comment by turambar** [comment]
@rohrl thank you for your persistence in testing model import. I'm traveling this week so I don't have a ton of dev cycles, but I'll try to take a look when I settle back down (weekend or early next week at latest). In the meantime, I encourage you to fire up your debugger and dive into the source code -- even if you just narrow it down some, that will help accelerate a fix, if one is needed.

@eraly if you get any bandwidth, maybe take a look? Testing Inception pretty thoroughly is on our post-release TODO list anyway, no?

**Comment by turambar** [comment]
Will take a look at this this week.

**Comment by maxpumperla** [comment]
@rohrl does this problem still occur in current version? Thanks
